v1/plugin/plugin.go

```diff
@@ -312,29 +507,212 @@ func startPlugin(srv server, m *meta, p *pluginProxy) int {
 	addr := fmt.Sprintf("127.0.0.1:%v", l.Addr().(*net.TCPAddr).Port)
 	lis, err := net.Listen("tcp", addr)
 	if err != nil {
-		// TODO(danielscottt): logging
-		panic(err)
+		return "", err
 	}
 	go func() {
 		err := srv.Serve(lis)
 		if err != nil {
-			panic(err)
+			log.Fatal(err)
 		}
 	}()
+	pprofAddr := "0"
+	if isPprof {
+		pprofAddr, err = startPprof()
+		if err != nil {
+			return "", err
+		}
+	}
 	resp := preamble{
 		Meta:          *m,
 		ListenAddress: addr,
 		Type:          m.Type,
-		PprofAddress:  pprofPort,
+		PprofAddress:  pprofAddr,
 		State:         0, // Hardcode success since panics on err
 	}
 	preambleJSON, err := json.Marshal(resp)
 	if err != nil {
 		panic(err)
 	}
-	libInputOutput.printOut(string(preambleJSON))
-	go p.HeartbeatWatch()
-	// TODO(danielscottt): exit code
-	<-p.halt
-	return 0
+
+	return string(preambleJSON), nil
+}
+
+// GetPluginType converts a pluginType to a string
+// describing what type of plugin this is
+func getPluginType(plType pluginType) string {
+	switch plType {
+	case collectorType:
+		return "collector"
+
+	case publisherType:
+		return "publisher"
+
+	case processorType:
+		return "processor"
+	}
+	return ""
+}
+
+// GetRPCType converts a RPCType (int) to a string
+// as described in snap/control/plugin/plugin.go
+func getRPCType(i int) string {
+	if i == 0 {
+		return "NativeRPC (0)"
+	} else if i == 2 {
+		return "GRPC (2)"
+	}
+	return "Not a valid RPC Type"
+}
+
+func showDiagnostics(m meta, p *pluginProxy, c Config) error {
+	defer timeTrack(time.Now(), "showDiagnostics")
+	printRuntimeDetails(m)
+	err := printConfigPolicy(p, c)
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	met, err := printMetricTypes(p, c)
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
```

Comment:
I have a concern of using `os.Exit(1)` here - it will be simple:
-  just return error or
- remove  "error" from the output and make simple `return` in lines 578 and 583 (based on fact the output of showDiagnostics is nowhere handled)